packages/dev-server-hmr/src/HmrPlugin.ts

```diff
@@ -0,0 +1,284 @@
+
+    webSockets.on('message', ({ webSocket, data }) => this._onMessage(webSocket, data));
+    fileWatcher.on('change', path => this._onFileChanged(path));
+    fileWatcher.on('unlink', path => this._onFileChanged(path));
+
+    this._logger?.debug('[hmr] Listening for HMR messages');
+  }
+
+  /** @inheritDoc */
+  async serve(context: Context) {
+    // Someone is requesting the injected client script
+    if (context.path === NAME_HMR_CLIENT_IMPORT) {
+      if (!this._webSockets) {
+        return;
+      }
+      return hmrClientScript(this._webSockets);
+    }
+
+    // We are serving a new file or it has changed, so clear all the
+    // dependencies we previously tracked (if any).
+    this._clearDependencies(context.path);
+    this._logger?.debug(`[hmr] Cleared dependency tree cache of ${context.path}`);
+  }
+
+  /** @inheritDoc */
+  async transformImport({ source, context }: { source: string; context: Context }) {
+    const importPath = pathUtil.resolve(context.path, source);
+
+    if (importPath === NAME_HMR_CLIENT_IMPORT) {
+      return;
+    }
+
+    const mod = this._getOrCreateModule(context.path);
+    const dependencyMod = this._getOrCreateModule(importPath);
+
+    mod.dependencies.add(importPath);
+    dependencyMod.dependents.add(context.path);
+    this._logger?.debug(`[hmr] Added dependency from ${context.path} -> ${importPath}`);
+  }
+
+  /** @inheritDoc */
+  async transform(context: Context) {
+    // If the module references import.meta.hot it can be assumed it
+    // supports hot reloading
+    const hmrEnabled = context.body.includes('import.meta.hot') === true;
+    const mod = this._getOrCreateModule(context.path);
+    mod.hmrEnabled = hmrEnabled;
+    this._logger?.debug(`[hmr] Setting hmrEnabled=${hmrEnabled} for ${context.path}`);
+
+    if (hmrEnabled && context.response.is('js')) {
+      return `
+        import {create as __WDS_HMR__} from '${NAME_HMR_CLIENT_IMPORT}';
+        import.meta.hot = __WDS_HMR__(import.meta.url);
+        ${context.body}
+      `;
+    }
+  }
+
+  /**
+   * Clears the dependency cache/tree of a particular module
+   * @param path Module path to clear
+   */
+  protected _clearDependencies(path: string): void {
+    const mod = this._getModule(path);
+
+    if (!mod) {
+      return;
+    }
+
+    for (const dep of mod.dependencies) {
+      const depMod = this._getModule(dep);
+      if (depMod) {
+        depMod.dependents.delete(path);
+      }
+    }
+
+    mod.dependencies = new Set();
+  }
+
+  /** @inheritDoc */
+  serverStop() {
+    this._webSockets = undefined;
+  }
+
+  /**
+   * Fired when a file has changed.
+   * @param path Module path which has changed
+   */
+  protected _onFileChanged(path: string): void {
+    // If we know what this module is, we can try trigger a HMR update
+    if (this._hasModule(path)) {
+      this._triggerUpdate(path);
+      return;
+    }
+
+    // Otherwise we reload the page
+    this._broadcast({ type: 'hmr:reload' });
```

Comment:
im not sure this is right... if a file changes which we've not yet requested, we will send a reload command.

but maybe a file changed for something we just didn't visit yet, so im wondering if this should just do nothing?

Reply:
could this be css file changes as well? 🤔
we should be able to "HMR" css files right? full replace css "text"

but potentially out of scope for this initial implementation 🤔

Reply:
There are no CSS modules